docs/design.md

```diff
@@ -0,0 +1,371 @@
+with low-level data repository files known as CARs (Content ARchives), on both
+the read and the write sides.
+
+Note that Badger is generally a fitting database for write-intense and
+iterator-heavy workloads. It just doesn't withstand the pattern of usage we
+subject it to in large IPLD-based software components, especially past the 100s
+GBs data volume mark.
+
+For more information on motivation and architecture, please refer to
+[CAR-native DAG store](https://docs.google.com/document/d/118fJdf8HGK8tHYOCQfMVURW9gTF738D0weX-hbG-BvY/edit#). This document is 
+recommended as a pre-read.
+
+## Overview
+
+The DAG store comprises three layers:
+
+1. Storage layer (manages shards)
+2. Index repository (manages indices)
+3. DAG access layer (manages queries)
+
+## Storage layer
+
+The DAG store holds shards of data. In Filecoin, shard = deal. Each shard is a
+repository of data capable of acting as a standalone blockstore.
+
+### Shards
+
+Shards are identified by an opaque byte string: the shard key. In the case of
+Filecoin, the shard key is the `PieceCID` of the storage deal.
+
+A shard contains:
+
+1. the shard key (identifier).
+2. the means to obtain shard data (a `Mount` that provides the CAR either
+   locally, remotely, or some other location).
+3. the shard index (both acting as a manifest of the contents of the shard, and
+   a means to efficiently perform random reads).
+
+### CAR formats
+
+A shard can be filled with CARv1 and CARv2 data. CARv2 can be indexed or
+indexless.
+
+The choice of version and characteristics affects how the shard index is
+populated:
+
+1. CARv1: the index is calculated upon shard activation.
+2. Indexless CARv2: the index is calculated upon shard activation.
+3. Indexed CARv2: the inline index is adopted as-is as the shard index.
+
+### Shard states
+
+1. **Available:** the system knows about this shard and is capable of serving
+   data from it instantaneously, because (a) an index exists and is loaded, and
+   (b) data is locally accessible (e.g. it doesn't need to be fetched from a
+   remote mount).
+2. **Unavailable:** the system knows about this shard, but is not capable of
+   serving data from it because the shard is being initialized, or the mount is
+   not available locally, but still accessible with work (e.g. fetching from a
+   remote location).
+3. **Destroyed**: the shard is no longer available; this is permanent condition.
+
+### Operations
+
+#### Shard registration
+
+To register a new shard, call:
+
+```go
+dagst.RegisterShard(key []byte, mount Mount, opts ...RegisterOpts) error
+```
+
+1. This method takes a shard key and a `Mount`.
+2. It initializes the shard in `Unavailable` state.
+3. Calls `mount.Info()` to determine if the mount is of local or remote type.
+4. Calls `mount.Stat()` to determine if the mount target exists. If not, it
+   errors the registration.
+5. If remote, it fetches the remote resource into the scrap area.
```

Comment:
IMO, users should have the option to do this lazily. For a use case like Lotus, we don't want to fetch a remote Unsealed sector file into the scrap area at Shard Registration time because the registration would be done when a storage deal is completed but the CAR file should be copied to the local scrap area only when a retrieval is requested.

But then, we do need to fetch the unsealed sector file at shard registration to Index it when a storage deal is done. In that case, I think the user call flow for a Lotus storage deal  would look like:

1. `dagst.RegisterShard`(fetch, Index and mark shard as available) followed by a:
2. `dagst. ReleaseShard` (release local scrap copy, mark shard as unavailable) .

Then, during  retrieval, the DAG access by the user would result into the DAG Store fetching the unsealed CAR file locally (scrap area copy).

Reply:
👍  Yes, this sounds correct.